src/edu/csus/ecs/pc2/ui/ShadowControlPane.java

```diff
@@ -698,7 +735,79 @@ public void run() {
         }
         return testConnectionButton;
     }
+
+    /*
+     * IShadowMonitorStatus implementaiton
+     */
+    /**
+     * {@inheritDoc}
+     */
+    public void updateShadowLastToken(String token)
+    {
+        // if the value supplied is valid and different from what we last saw,
+        // then we update the text fields
+        if(token != null) {
+            if(lastToken == null || !token.equals(lastToken)) {
+                // TODO: Do we want to save the token to a file here in case we crash?
+                //       Currently, token is only saved when the shadow is "stopped"
+                // TODO: Do we want to "InvokeLater" these (simple) updates to text fields?
+                lastToken = token;
+                lastEventTextfield.setText(lastToken);
+                try {
+                    GregorianCalendar cal = new GregorianCalendar();
+                    
+                    lastDateFormat.setCalendar(cal);
+                    lastEventTimeTextField.setText(lastDateFormat.format(cal.getTime()));
+                } catch(Exception e) {
+                    // Just ignore any exception from date formatter
+                }
+            }
+        }
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    public void updateShadowNumberofRecords(int nRec)
+    {
+        // if the number of records is different from what we last display and it's valid
+        // update the instrumentation.
+        if(nRec != numRecord && nRec >= 0) {
+            // TODO: Do we want to "InvokeLater" this (simple) update to a text field?
+            numRecord = nRec;
+            // Save to file? Send to server contestinfo?
+            lastRecordTextfield.setText(String.valueOf(numRecord));
+        }
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    public void connectFailed(String token)
+    {
+        // Update JTable - future commit
```

Comment:
Add a // TODO comment or // SOMEDAY comment to make it easier to identify this future work

Reply:
Updated comments to show TODO